fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/storestate/MetaDataVersionStampStoreStateCache.java

```
@@ -0,0 +1,143 @@
+ *
+ * @see FDBRecordStore#setStateCacheabilityAsync(boolean)
+ * @see FDBRecordContext#getMetaDataVersionStamp(IsolationLevel)
+ */
+public class MetaDataVersionStampStoreStateCache implements FDBRecordStoreStateCache {
+    @Nonnull
+    private final FDBDatabase database;
+    @Nonnull
+    private final Cache<SubspaceProvider, FDBRecordStoreStateCacheEntry> cache;
+
+    MetaDataVersionStampStoreStateCache(@Nonnull FDBDatabase database, @Nonnull Cache<SubspaceProvider, FDBRecordStoreStateCacheEntry> cache) {
+        this.database = database;
+        this.cache = cache;
+    }
+
+    @Nonnull
+    private FDBRecordStoreStateCacheEntry getNewerEntry(@Nonnull FDBRecordStoreStateCacheEntry entry1, @Nonnull FDBRecordStoreStateCacheEntry entry2) {
+        if (entry1.getMetaDataVersionStamp() == null) {
+            return entry2;
+        } else if (entry2.getMetaDataVersionStamp() == null) {
+            return entry1;
+        } else {
+            return ByteArrayUtil.compareUnsigned(entry1.getMetaDataVersionStamp(), entry2.getMetaDataVersionStamp()) >= 0 ? entry1 : entry2;
+        }
+    }
+
+    private void addToCache(@Nonnull SubspaceProvider subspaceProvider, @Nonnull FDBRecordStoreStateCacheEntry cacheEntry) {
+        cache.asMap().merge(subspaceProvider, cacheEntry, (entry1, entry2) -> {
+            final FDBRecordStoreStateCacheEntry newerEntry = getNewerEntry(entry1, entry2);
+            if (newerEntry.getRecordStoreState().getStoreHeader().getCacheable()) {
+                return newerEntry;
+            } else {
+                return null;
+            }
+        });
+    }
+
+    private void invalidateOlderEntry(@Nonnull SubspaceProvider subspaceProvider, @Nonnull byte[] metaDataVersionStamp) {
+        cache.asMap().computeIfPresent(subspaceProvider, (ignore, existingEntry) -> {
+            // Invalidate the key unless the cached meta-data is newer than this meta-data version stamp
```

It's actually newer than or equal to, but I guess equal is impossible with proper usage, isn't it?

---

I updated the comment to match the code.

I think that that's right--that with proper usage, the two would never be equal. In particular, this is only called by someone who thinks that for `metaDataVersionStamp = x`, that the store state isn't cacheable, so if the existing entry had a value equal to `metaDataVersionStamp` and *was* cached, that would mean that the cache has two different ideas as to the state of the meta-data at that meta-data version stamp.

I suppose, in theory, that should throw an error?